src/api/customization.rs

```diff
@@ -50,27 +88,110 @@ pub trait Customization {
+    /// - CTAP2.1: Maximum PIN retries must be 8 at most.
+    ///
+    /// The fail retry counter is reset after entering the correct PIN.
+    fn max_pin_retries(&self) -> u8;
+
+    /// Enables or disables signature counters.
+    ///
+    /// The signature counter is currently implemented as a global counter.
+    /// The specification strongly suggests to have per-credential counters.
+    /// Implementing those means you can't have an infinite amount of server-side
+    /// credentials anymore. Also, since counters need frequent writes on the
+    /// persistent storage, we might need a flash friendly implementation. This
+    /// solution is a compromise to be compatible with U2F and not wasting storage.
+    ///
+    /// https://www.w3.org/TR/webauthn/#signature-counter
+    fn use_signature_counter(&self) -> bool;
+
+    // ###########################################################################
+    // Constants for performance optimization or adapting to different hardware.
+    //
+    // Those constants may be modified before compilation to tune the behavior of
+    // the key.
+    // ###########################################################################
+
+    /// Limits the number of RP IDs that can change the minimum PIN length.
+    ///
+    /// # Invariant
+    ///
+    /// - If this value is 0, default_min_pin_length_rp_ids() must be non-empty.
+    ///
+    /// You can use this constant to have an upper limit in storage requirements.
+    /// This might be useful if you want to more reliably predict the remaining
+    /// storage. Stored string can still be of arbitrary length though, until RP ID
+    /// truncation is implemented.
+    /// Outside of memory considerations, you can set this value to 0 if only RP IDs
+    /// in default_min_pin_length_rp_ids() should be allowed to change the minimum
+    /// PIN length.
+    fn max_rp_ids_length(&self) -> usize;
 }
 
 #[derive(Clone)]
 pub struct CustomizationImpl {
     pub default_cred_protect: Option<CredentialProtectionPolicy>,
+    pub default_min_pin_length: u8,
+    pub default_min_pin_length_rp_ids: &'static [&'static str],
+    pub enforce_always_uv: bool,
     pub max_msg_size: usize,
+    pub max_pin_retries: u8,
+    pub use_signature_counter: bool,
+    pub max_rp_ids_length: usize,
 }
 
 pub const DEFAULT_CUSTOMIZATION: CustomizationImpl = CustomizationImpl {
+    default_min_pin_length: 4,
+    default_min_pin_length_rp_ids: &[],
+    enforce_always_uv: false,
     default_cred_protect: None,
     max_msg_size: 7609,
+    max_pin_retries: 8,
+    use_signature_counter: true,
+    max_rp_ids_length: 8,
 };
 
 impl Customization for CustomizationImpl {
     fn default_cred_protect(&self) -> Option<CredentialProtectionPolicy> {
         self.default_cred_protect
     }
 
+    fn default_min_pin_length(&self) -> u8 {
+        self.default_min_pin_length
+    }
+
+    fn default_min_pin_length_rp_ids(&self) -> Vec<String> {
```

Comment:
@ia0 Now we do this, so this customization (and ENTERPRISE_RP_ID_LIST) won't be treated as constants anymore. But I think it should be fine? There's very limited code path affected.

Reply:
Indeed they look less like constants. But I would be a bit surprised if some optimizations would not trigger. The optimization we rely on is when we iterate on this list in `set_min_pin_length_rp_ids`. The whole loop should be eliminated. We could try to check that it's still the case. One possible trick is to call an extern function in the loop and see if the linker complains:

```rust
    for &rp_id in DEFAULT_MIN_PIN_LENGTH_RP_IDS.iter() {
        // Add the 2 following lines to test if it still links
        extern "C" { fn should_not_be_called(); }
        unsafe { should_not_be_called() };
        let rp_id = String::from(rp_id);
        if !min_pin_length_rp_ids.contains(&rp_id) {
            min_pin_length_rp_ids.push(rp_id);
        }
    }
```

Reply:
The optimization won't trigger. I think it's because to eliminate the loop code the function has be able to be computed at compilation time. But we have some heap allocation like `String::from` in that function which make it impossible to do so?

Reply:
I don't think the `String::from` should matter if the list is empty. But yeah, it's possible the constant propagation is not aggressive enough. In particular, if `collect` is not able to see that the iterator is empty, then we won't get the optimization.